### ISSUE TITLE
ESP32: Micropython 1.19.1 Setting DHCP Hostname leads to Guru Medition Error

### ISSUE THREAD

**Issue by chrisdecker1201** [micropython/micropython#8792]
I'm currently trying to update my project from Micropython 1.18 to 1.19.1.

I've an issue setting the DHCP Hostname in 1.19.1. I've no issues on 1.18.

Here steps to reproduce.

1. Create a `network.json` file and change SSID and PSK to your WLAN

~~~json
{"SSID": "YOURWIFISSID", "PSK": "YOURWIFIPASSWORD"}
~~~

2. Create a `main.py` file

~~~python
import network
from ujson import loads

class NetworkHandler():
    """class handling network connections"""

    def __init__(self, hostname, file: str = "network.config"):
        """initialize network handler"""
        self.file = file
        self.hostname = hostname
        self.config = {}
        if not isinstance(self.file, str):
            self.config = self.file
        else:
            try:
                with open(self.file, "r", encoding="utf-8") as json_config:
                    self.config = loads(json_config.read())
            except OSError as os_error:
                if os_error.errno != 2:
                    raise

    def connect_wifi(self):
        ssid = self.config["SSID"]
        psk = self.config["PSK"]
        wlan = network.WLAN(network.STA_IF)
        wlan.active(True)
        wlan.config(dhcp_hostname=self.hostname)
        if not wlan.isconnected():
            wlan = network.WLAN(network.STA_IF)
            wlan.active(True)
            wlan.connect(ssid, psk)

if __name__ == '__main__':
    network_handler = NetworkHandler("timeslime")
    network_handler.connect_wifi()
~~~

3. Execute `main.py` with mpremote

~~~bash
mpremote u0 mount . run main.py
~~~

4. Getting the following output

~~~log
Local directory . is mounted at /remote
/home/micropython/esp-idf-v4.2/components/freertos/queue.c:7Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x4000127a  PS      : 0x00060530  A0      : 0x8013efee  A1      : 0x3ffcce10  
A2      : 0x3f41dbbc  A3      : 0x00000000  A4      : 0x3ffce24c  A5      : 0x00000001  
A6      : 0x00000000  A7      : 0x3ffd5b24  A8      : 0x00000057  A9      : 0x3ffccdc0  
A10     : 0x00000001  A11     : 0x00000000  A12     : 0x3ffd5b60  A13     : 0x00000000  
A14     : 0x00000001  A15     : 0x3ffd5b38  SAR     : 0x00000018  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000000  LBEG    : 0x4000c28c  LEND    : 0x4000c296  LCOUNT  : 0x00000000  

Backtrace:0x40001277:0x3ffcce10 0x4013efeb:0x3ffcce20 0x4013fc55:0x3ffcce40 0x4013e7ef:0x3ffcce70 0x4013ec40:0x3ffccea0 0x400f94c2:0x3ffccec0 0x400dbdc5:0x3ffccf80 0x400e2679:0x3ffccfb0 0x400e27a9:0x3ffccfd0 0x40084789:0x3ffccff0 0x400dbd66:0x3ffcd090 0x400e2679:0x3ffcd100 0x400e27a9:0x3ffcd120 0x40084789:0x3ffcd140 0x400dbd66:0x3ffcd1e0 0x400e2679:0x3ffcd230 0x400e26a2:0x3ffcd250 0x400efbca:0x3ffcd270 0x400efd5d:0x3ffcd300 0x400d4d12:0x3ffcd350


ELF file SHA256: 00495403985fb7e0

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12344
ho 0 tail 12 room 4
load:0x40080400,len:4124
entry 0x40080680
MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
Type "help()" for more information.
>>> 
~~~

5. When is comment this line, I've no issues and everything works fine.

~~~python
#wlan.config(dhcp_hostname=self.hostname)
~~~

I'm happy to help with more informations. If someone can explain me how to debug such an issue, I'm happy to learn to find the issue by my self.

**Comment by dpgeorge** [comment]
What hardware do you have?  What IDF version?  Do this at the REPL:
```
import sys
sys.version
sys.implementation
import platform
platform.platform()
```

**Comment by chrisdecker1201** [comment]
I have a ESP32 DevKit ESP32-WROOM Wireless WiFi and Bluetooth Developerboard

~~~log
import sys
>>> sys.version
'3.4.0; MicroPython v1.19.1 on 2022-06-18'
>>> sys.implementation
(name='micropython', version=(1, 19, 1), _machine='ESP32 module with ESP32', _mpy=10246)
>>> import platform
>>> platform.platform()
'MicroPython-1.19.1-xtensa-IDFv4.2.2-with-newlib3.0.0'
~~~

Is this enough information?

**Comment by dpgeorge** [comment]
Thanks for the info.

I can reproduce the problem, using IDFv4.4.1 and the GENERIC build running on a TinyPICO.

Inserting a `time.sleep_us(100)` between the `wlan.active(True)` and the `wlan.config(dhcp_hostname=self.hostname)` seems to fix it.  This could point to a race condition in the IDF, that it doesn't like configuring the DHCP hostname while the interface is coming up.

To work around the problem for now, I'd suggest adding that small delay.

**Comment by chrisdecker1201** [comment]
@dpgeorge Thank you very much for your help.

Should we create a bug in the IDF repository or is this an issue micropython can fix?

**Comment by IhorNehrutsa** [comment]
@chrisdecker1201 
I find this strange.
If I understand correctly the "timeslime" is dhcp server name.
I.e. you know that your local network has dhcp server "timeslime".
Do you want to get esp32 ip configuration from known for you predefined dhcp server?
In other case you should comment line:
```
# wlan.config(dhcp_hostname=self.hostname)
```
In common case you don't know dhcp server name.

**Comment by chrisdecker1201** [comment]
@IhorNehrutsa With `dhcp_hostname` you can set the hostname for your device. "timeslime" is than the name I see in the router.

**Comment by chrisdecker1201** [comment]
But then you've also to update the documentation

https://docs.micropython.org/en/latest/library/network.WLAN.html?highlight=dhcp

Following are commonly supported parameters (availability of a specific parameter depends on network technology type, driver, and [MicroPython port](https://docs.micropython.org/en/latest/reference/glossary.html#term-MicroPython-port)).

Parameter | Description
-- | --
dhcp_hostname | The DHCP hostname to use

**Comment by chrisdecker1201** [comment]
@dpgeorge I'm not sure, if I've the time and knowledge to provide a C example. Can someone else provide a C example? Or maybe can they (owners of the IDF repo) work with the micropython example I provided here?

**Comment by RSC-Games** [comment]
I noticed the crash behavior stops when you start the WLAN in AP mode. Is there a reason for this?

**Comment by dpgeorge** [comment]
This is hopefully fixed (at least worked around) by d6bc34a13aa734d8b32e5768c021377ac4815029, which adds a 1ms delay after activating the WLAN.

**Comment by RSC-Games** [comment]
Wow I forgot this issue existed.
